src/Orleans/CodeGeneration/GenericMethodInvoker.cs

```diff
@@ -0,0 +1,164 @@
+
+            // Load every argument from the argument array.
+            var methodParameters = concreteMethod.GetParameters();
+            for (var i = 0; i < methodParameters.Length; i++)
+            {
+                il.LoadArgument(1); // Load the argument array.
+
+                // The argument is offset by the type parameter count, since type parameters come first.
+                il.LoadConstant(i + this.typeParameterCount); 
+                il.LoadReferenceElement();
+                il.CastOrUnbox(methodParameters[i].ParameterType);
+            }
+
+            // Call the target method.
+            il.Call(concreteMethod);
+
+            // If the result type is Task or Task<T>, convert it to Task<object>.
+            var returnType = concreteMethod.ReturnType;
+            if (returnType != typeof(Task<object>))
+            {
+                var boxMethod = BoxMethods.GetOrAdd(returnType, CreateBoxMethod);
+                il.Call(boxMethod);
+            }
+
+            il.Return();
+            return il.CreateDelegate();
+        }
+
+        private static MethodInfo GetBoxMethod(Type returnType)
+        {
+            if (returnType == typeof(Task)) return TypeUtils.Method((Task task) => task.Box());
+            if (returnType == typeof(void)) return TypeUtils.Property(() => TaskDone.Done).GetMethod;
+
+            if (returnType.GetGenericTypeDefinition() != typeof(Task<>))
+                throw new ArgumentException($"Unsupported return type {returnType}.");
+            var innerType = returnType.GenericTypeArguments[0];
+            var methods = typeof(PublicOrleansTaskExtensions).GetMethods(BindingFlags.Static | BindingFlags.Public);
+            foreach (var method in methods)
+            {
+                if (method.Name != "Box" || !method.ContainsGenericParameters) continue;
+                return method.MakeGenericMethod(innerType);
+            }
+
+            throw new ArgumentException($"Could not find Box method for type {returnType}");
+        }
+
+        bool IEqualityComparer<object[]>.Equals(object[] x, object[] y)
+        {
+            if (ReferenceEquals(x, y)) return true;
+            if (ReferenceEquals(x, null)) return false;
+            if (ReferenceEquals(null, y)) return false;
+
+            if (x.Length < this.typeParameterCount || y.Length < this.typeParameterCount) return false;
+
+            for (var i = 0; i < this.typeParameterCount; i++)
+            {
+                if (!CompareAsTypes(x[i] as Type, y[i] as Type)) return false;
+            }
+
+            return true;
+        }
+
+        int IEqualityComparer<object[]>.GetHashCode(object[] obj)
+        {
+            if (obj == null || obj.Length == 0) return 0;
+            unchecked
+            {
+                // Only consider the type parameters.
+                var result = 0;
+                for (var i = 0; i < this.typeParameterCount && i < obj.Length; i++)
```

Comment:
If we included other values in our comparison, the hash code would be different for almost every call to a method, so we would generate IL very frequently. The goal is to generate IL once for each variant of a generic method.